### PR TITLE
Improve build system and library handling

### PR DESCRIPTION
**Pull request by GPUCode**
This PR is mostly comprised from yuzu backports that improve the current build system, mostly though to address issue #6204
Citra now searches for most required packages with find_package before falling back the external repos. I've also included some other minor fixes I found worthwhile from yuzu.

Things to do before merging
- [ ] Provide more explicit library control
- [ ] Make use of FetchContent_MakeAvailable and find_package integration
- [ ] Fix build errors on all platforms

- - -
This change is [Reviewable](https://reviewable.io/reviews/citra-emu/citra/6217)

### CODE REVIEW

**Comment by ImperatorStorm** [comment]
Partially related: https://github.com/citra-emu/citra/issues/5889

File: externals/find-modules/FindDiscordRPC.cmake [hunk 1]
```diff
@@ -0,0 +1,27 @@
+# SPDX-FileCopyrightText: 2022 Alexandre Bouvier <contact@amb.tf>
+#
+# SPDX-License-Identifier: GPL-3.0-or-later
```
Comment (SachinVin) [inline externals/find-modules/FindDiscordRPC.cmake:3]: This is only part of the build system, but the licence potentially conflicting with ours.

File: externals/CMakeLists.txt [hunk 2]
```diff
@@ -125,24 +161,39 @@ if (ENABLE_WEB_SERVICE)
...
     endif()
+endif()
 
-    # httplib
-    add_library(httplib INTERFACE)
-    target_include_directories(httplib INTERFACE ./httplib)
-    target_compile_options(httplib INTERFACE -DCPPHTTPLIB_OPENSSL_SUPPORT)
-    target_link_libraries(httplib INTERFACE ${OPENSSL_LIBRARIES})
+# JSON
+if (ENABLE_WEB_SERVICE AND NOT TARGET nlohmann_json::nlohmann_json)
```
Comment (SachinVin) [inline externals/CMakeLists.txt:179]: Why remove these out of the `if(ENABLE_WEB_SERVICE)`, but keep openssl and httplib within them?

File: CMakeLists.txt [hunk 3]
```diff
@@ -20,7 +24,7 @@ endif()
 option(ENABLE_LTO "Enable link time optimization" OFF)
 
 option(ENABLE_SDL2 "Enable the SDL2 frontend" ON)
-option(USE_SYSTEM_SDL2 "Use the system SDL2 lib (instead of the bundled one)" OFF)
+option(CITRA_USE_SYSTEM_SDL2 "Use the system SDL2 lib (instead of the bundled one)" OFF)
```
Comment (SachinVin) [inline CMakeLists.txt:27]: Prefixing CITRA_ to only some options is very inconsistent.

File: src/input_common/CMakeLists.txt [hunk 4]
```diff
@@ -33,7 +33,5 @@ if(ENABLE_SDL2)
 endif()
 
 create_target_directory_groups(input_common)
-target_link_libraries(input_common PUBLIC core PRIVATE common ${Boost_LIBRARIES})
-target_include_directories(input_common PRIVATE ${LIBUSB_INCLUDE_DIR})
-target_link_libraries(input_common PUBLIC ${LIBUSB_LIBRARIES})
+target_link_libraries(input_common PUBLIC core PRIVATE common ${Boost_LIBRARIES} libusb::usb)
```
Comment (SachinVin) [inline src/input_common/CMakeLists.txt:36]: Nit, Boost and libusb should be in a different `target_link_libraries`, so it is consistent with the rest of the codebase

File: src/core/CMakeLists.txt [hunk 5]
```diff
@@ -476,12 +476,13 @@ endif()
 create_target_directory_groups(core)
 
 target_link_libraries(core PUBLIC common PRIVATE audio_core network video_core)
-target_link_libraries(core PUBLIC Boost::boost PRIVATE cryptopp fmt::fmt open_source_archives Boost::serialization)
+target_link_libraries(core PUBLIC Boost::boost Boost::serialization)
```
Comment (SachinVin) [inline src/core/CMakeLists.txt:479]: Q. And maybe an exercise for another day, Does boost stuff need to be public here?

File: externals/libusb/CMakeLists.txt [hunk 6]
```diff
@@ -150,3 +150,4 @@ check_include_files(sys/timerfd.h USBI_TIMERFD_AVAILABLE)
 
 
 configure_file(config.h.in config.h)
+add_library(libusb::usb ALIAS usb)
```
Comment (SachinVin) [inline externals/libusb/CMakeLists.txt:153]: Nit, new line

**Comment by florhizome** [comment]
ping ;)

File: externals/CMakeLists.txt (hunk 2 above)
Comment (GPUCode) [reply externals/CMakeLists.txt:179]: Each library gets it's own if. httblib depends on openssl so they are in the same if. yuzu does the same here https://github.com/yuzu-emu/yuzu/blob/9083ad816fc0fca2e39c8c526ae85e3d09db0229/externals/CMakeLists.txt#L104 However I can merge them if desired

File: externals/find-modules/FindDiscordRPC.cmake (hunk 1 above)
Comment (GPUCode) [reply externals/find-modules/FindDiscordRPC.cmake:3]: Can we ask the yuzu developer to re-license this or provide an exemption?

File: CMakeLists.txt (hunk 3 above)
Comment (GPUCode) [reply CMakeLists.txt:27]: What do you recommend, using CITRA_ everywhere or dropping this? For the question the answer is no, dynarmic also recently dropped the BUNDLE_ variables https://github.com/merryhime/dynarmic/pull/725. It's a good idea to always prefer system libraries unless unavailable. However we can maintain a list of excluded libraries similar to the architectures macro

File: src/core/CMakeLists.txt (hunk 5 above)
Comment (GPUCode) [reply src/core/CMakeLists.txt]: I tested it and it can be made PRIVATE

File: CMakeLists.txt (hunk 3 above)
Comment (SachinVin) [reply CMakeLists.txt:25]: I think we can just drop the `USE_SYSTEM_SDL2` and friends, altogether in favour of the cmake provided option. Also, dynarmic still allows you to disable all system libraries, https://github.com/merryhime/dynarmic/blob/master/CMakeLists.txt#L123-L130

File: externals/CMakeLists.txt (hunk 2 above)
Comment (SachinVin) [reply externals/CMakeLists.txt:200]: I think it makes it easier to read if they are all in a single if

**Comment by ImperatorStorm** [comment]
ping, has a couple conflicts.

**Pull request closed without merging**